### PR TITLE
#260 & #247 stop messing up files

### PR DESCRIPTION
**Pull request by sinopsysHK**
Using a hash function with lower collision over the whole file name prevents from confusing between a new file and an older one that would have had the same hash.

Current hash was computed upon the first 16 chars of the file name so if ever there is a common prefix (such as "[myfevoriteindexer.com]")  the hash  is the same.

Now with the full file name being used it reduce the risk of collision which will avoid to retrieve old data from previous run (#247 )

Furthermose having the same guid convention in postprocessing (lower case) it enables to find back a previous processing and avoid seeking again for movie metadata when files are not deleted from download directory (#260)

Added little optimization avoiding blank tmdb queries.

Added a patch to secure Kodi import (movies with no video details info crashing import).

### CODE REVIEW

**Comment by nosmokingbandit** [comment]
There is a lot going on here...

See my comments on the file changes.

File: core/ajax.py [hunk 1]
```diff
@@ -871,6 +871,11 @@ def import_kodi_movies(self, movies):
 
         for movie in movies:
 
+            if not movie['imdbid']:
```
Comment (nosmokingbandit) [inline core/ajax.py]: If Kodi doesn't provide the imdbid there should be a means for the user to provide this in the webui in the way that Plex and Directory imports do. Simply ignoring a missing imdbid is not a good idea.

File: core/library.py [hunk 2]
```diff
@@ -157,7 +158,7 @@ def get_movies(url):
                 else:
                     movie['resolution'] = 'DVD-SD'
             else:
-                movie['videocodec'] = movie['resolution'] = ''
+                movie['videocodec'] = ''
```
Comment (nosmokingbandit) [inline core/library.py]: If you remove this the 'resolution' key will not exist if the metadata cannot be parsed by hachoir. This key should be assigned here so that it is a blank string rather than null in the database.

File: core/library.py [hunk 3]
```diff
@@ -908,7 +909,7 @@ def searchresults(guid, status, movie_info=None):
 
             search_result = searchresults.score([search_result], imported=True)[0]
 
-            required_keys = ('score', 'size', 'status', 'pubdate', 'title', 'imdbid', 'indexer', 'date_found', 'info_link', 'guid', 'torrentfile', 'resoluion', 'type', 'downloadid', 'freeleech')
+            required_keys = ('score', 'size', 'status', 'pubdate', 'title', 'imdbid', 'indexer', 'date_found', 'info_link', 'guid', 'torrentfile', 'resolution', 'type', 'downloadid', 'freeleech')
```
Comment (nosmokingbandit) [inline core/library.py]: If you go looking for typos you'll quickly be overwhelmed. Too bad IDEs don't have good spell checkers.

File: core/scheduler.py [hunk 4]
```diff
@@ -116,7 +116,7 @@ def scan_directory():
             if r:
                 logging.info('Found match for {} in releases: {}.'.format(fname, r['title']))
             else:
-                r['guid'] = 'POSTPROCESSING{}'.format(b16encode(fname.encode('ascii', errors='ignore')).decode('utf-8').zfill(16)[:16])
+                r['guid'] = 'postprocessing{}'.format(hashlib.md5(fname.encode('ascii', errors='ignore')).hexdigest())
```
Comment (nosmokingbandit) [inline core/scheduler.py]: I assumed b16 would be faster but after a quick test it looks like md5 is slightly faster. Both call C functions for the actual work but I suppose the md5 algorithm is so optimized at this point in time.

File: core/ajax.py (hunk 1 above)
Comment (nosmokingbandit) [inline core/ajax.py]: If Kodi doesn't provide an imdbid we should give the user a chance to supply it in the webui similar to how the Plex and Directory import pages handle missing data.

File: core/library.py (hunk 2 above)
Comment (nosmokingbandit) [inline core/library.py]: The resolution key needs to be assigned somewhere or it will be null in the database (None in python) and will raise exceptions any time it is compared to other strings.

File: core/library.py (hunk 3 above)
Comment (nosmokingbandit) [inline core/library.py]: Don't go looking for typos, you'll quickly be overwhelmed. Too bad IDEs don't have good spell checkers.

Comment (sinopsysHK) [reply core/library.py]: Sure typo may not be the priority but in this case it has an impact on what is stored in the DB and thus loose info that is never recovered later on

File: core/scheduler.py (hunk 4 above)
Comment (sinopsysHK) [reply core/scheduler.py]: I understand the focus on performance but there is also a need to be correct: very frustrating to play a movie that is not actually what its filename tells it is.
Furthermore as mover/renamer overwrite existing files without warning it can mess the current library.

As b16encode do not apply any rotate/swap of bytes then it is not a good hash function.

File: core/ajax.py (hunk 1 above)
Comment (sinopsysHK) [reply core/ajax.py]: Understand but do changing behaviour in ui prevents the need to control there is actually something to query tmdb for ?
You might have personal videos that do not have IMDb Id: it is not a missing metadata from Kodi search.

But it indeed it would be good to have mean to hand over the process and found missing ones.

File: core/scheduler.py (hunk 4 above)
Comment (nosmokingbandit) [reply core/scheduler.py]: The purpose is simply to have a unique string in the database as a fake guid. Using md5 or b16 doesn't make any difference in how the string is then used since an identical directory or filename will result in an identical guid. There is no real need to obscure the data at all, I simply wanted to have a guid-like representation.

Changing this to md5 will have no affect on anything other than performance. Which is actually slightly better than b16 for whatever reason.

File: core/library.py (hunk 3 above)
Comment (nosmokingbandit) [reply core/library.py]: I'm not saying it doesn't need to be fixed, only that finding all of them would take a full day.

And its *lose*.

File: core/library.py (hunk 2 above)
Comment (sinopsysHK) [reply core/library.py]: “Unknown” is even more accurate but I wanted to respect the pattern and further more Unknown-SD is a used else where so integration was closer to as-is.

**Comment by sinopsysHK** [comment]
To wrap up all the reviews, except if I miss understood, I feel like you agreed with all changes including Ajax.py to support specific Kodi use-case where library can contain personal video with no IMDb Id and library.py that keeps resolution attribute set in any case.
Are you still expecting some back tracking ?

**Comment by nosmokingbandit** [comment]
I think the only thing I want changed before I merge is switching `Unknown-SD' to `Unknown`. I don't want to indicate something is SD if we really don't know. Some users have bandwidth caps and I don't want them to accidentally download a massive 4K release because of a parsing error that caused it to default to SD. It won't change how anything actually works, but makes it a bit more clear to the user that Unknown means that we have no idea at all and they are assuming all risk associated with that.
